utils/treesitter/grammar.js

```diff
@@ -0,0 +1,507 @@
+          prec(
+            precedence,
+            seq(field('operator', operator), field('value', $._expression))
+          )
+        )
+      );
+    },
+
+    binary_expression: ($) => {
+      const table = [
+        [PREC.LogicalAnd, 'and'],
+        [PREC.LogicalOr, 'or'],
+        [PREC.BitwiseAnd, '&'],
+        [PREC.BitwiseOr, '|'],
+        [PREC.BitwiseXor, '^'],
+        [PREC.BitShift, choice('<<', '>>')],
+        [PREC.Relational, choice('==', '!=', '<', '<=', '>', '>=')],
+        [PREC.Additive, choice('+', '-')],
+        [PREC.Multiplicative, choice($.binary_star, '/', '%')],
+      ];
+
+      return choice(
+        ...table.map(([precedence, operator]) =>
+          prec.left(
+            precedence,
+            seq(
+              field('left', $._expression),
+              field('operator', operator),
+              field('right', $._expression)
+            )
+          )
+        )
+      );
+    },
+
+    // This should be non-associative but conflicts are not allowed in tree-sitter
+    as_expression: ($) => prec.left(seq($._expression, 'as', $._expression)),
+
+    ref_expression: ($) => prec.right(PREC.TermPrefix, seq('&', $._expression)),
+
+    deref_expression: ($) =>
+      prec.right(PREC.TermPrefix, seq('*', $._expression)),
+
+    fn_type_expression: ($) =>
+      prec.left(seq('__Fn', $.tuple, '->', $._expression)),
+
+    if_expression: ($) =>
+      prec(
+        PREC.IfExpression,
+        seq('if', $._expression, 'then', $._expression, 'else', $._expression)
+      ),
+
+    paren_expression: ($) => seq('(', comma_sep($._expression), ')'),
+
+    tuple: ($) => $.paren_expression,
+
+    index_expression: ($) =>
+      prec(PREC.TermPostfix, seq($._expression, '[', $._expression, ']')),
+
+    designator: ($) => seq('.', choice('base', $.ident)),
+
+    postfix_expression: ($) =>
+      prec(
+        PREC.TermPostfix,
+        seq(
+          $._expression,
+          choice(
+            '++',
+            '--',
+            $.designator,
+            seq('->', $.ident),
+            seq(choice('.', '->'), '(', $._expression, ')')
+          )
+        )
+      ),
+
+    where_clause: ($) =>
+      prec(
+        PREC.WhereClause,
+        choice(
+          seq($._expression, '==', $._expression),
+          seq($._expression, 'impls', $._expression),
+          seq($.designator, '=', $._expression),
+          prec.left(seq($.where_clause, 'and', $.where_clause))
+        )
+      ),
+
+    where_expression: ($) =>
+      prec.left(PREC.TermPostfix, seq($._expression, 'where', $.where_clause)),
+
+    call_expression: ($) => prec(PREC.TermPostfix, seq($._expression, $.tuple)),
+
+    pointer_expression: ($) =>
+      prec(PREC.TypePostfix, seq($._expression, $.postfix_star)),
+
+    _expression: ($) =>
+      choice(
+        $.ident,
+        $.pointer_expression,
+        $.index_expression,
+        $.fn_type_expression,
+        $.designator,
```

Comment:
Is this needed? I don't think designators are generally allowed in expressions at the moment.